### PR TITLE
extmod/uzlib: Add gzip compression support. (update to #5613)

### PR DESCRIPTION
**Pull request by harbaum**
This is a new attempt on https://github.com/micropython/micropython/pull/5613. It addresses all of the comments and additionally fixes a memory leak.

This has been stress tested with a few GB of video data as seen [here](https://www.youtube.com/watch?v=9zsAbTVxjMw).

As a quick test gzip compression can be tested against the already present decompression like so:

```
>>> uzlib.decompress(uzlib.gzip("Hello World!")[10:-8], -15)
compressed from 12 to 14 raw bytes
bytearray(b'Hello World!')
```

### CODE REVIEW

**Review by andrewleech — CHANGES_REQUESTED** [review]

File: extmod/uzlib/defl_static.c [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+#include <stdlib.h>
+#include <stdint.h>
+#include <string.h>
+#include <assert.h>
+#include "defl_static.h"
+
+#define snew(type) ( (type *) malloc(sizeof(type)) )
+#define snewn(n, type) ( (type *) malloc((n) * sizeof(type)) )
+#define sresize(x, n, type) ( (type *) realloc((x), (n) * sizeof(type)) )
+#define sfree(x) ( free((x)) )
```
[the remark this answers is not in the document]
Comment (harbaum) [reply extmod/uzlib/defl_static.c]: Done and verified that it still works

File: extmod/moduzlib.c [hunk 2]
```diff
@@ -206,9 +208,66 @@ STATIC mp_obj_t mod_uzlib_decompress(size_t n_args, const mp_obj_t *args) {
...
+    
+    // append 32 bit crc of original data
+    uint32_t offset = sizeof(gzip_header)+comp->out.outlen;
+    uint32_t crc = ~uzlib_crc32(bufinfo.buf, len, ~0);
+    memcpy(dest_buf+offset, &crc, sizeof(crc));
+    // append 32 bit length of original data
+    memcpy(dest_buf+offset+sizeof(crc), &len, sizeof(len));
+
+    // free all temporarily used memory
+    free(comp->out.outbuf);     // free internal buffer allocated by compression
```
[the remark this answers is not in the document]
Comment (harbaum) [reply extmod/moduzlib.c]: Done and verified that it still works

**Comment by dmazzella** [comment]
progress on this?

**Comment by harbaum** [comment]
Closing since files have moved and changed and the whole uzlib library needed an update.

New pull request at https://github.com/micropython/micropython/pull/8195

**Pull request closed without merging**